### PR TITLE
Don't try to override command line option -shader-cache-file-dir

### PR DESCRIPTION
**Pull request by jayfoad**
Without this patch I get this error, which doesn't make any sense:
$ amdllpc -shader-cache-file-dir=. /dev/null
amdllpc: for the --shader-cache-file-dir option: may only occur zero or one times!

I noticed this when cut and pasting all the command line options from
the "Compiler Options" line in a dumped .pipe file.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit d9cd7189824f2ec2ecfe747e55bf1768266755c4
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 9fab16015e522fff05890a045a1e9d8d3c23a636</li>
<li>SPVGEN: dev a61db14f75728e91c243d4bfc1c39700e9dc0ba6</li>
<li>XGL: dev 052ebfa2e77abd14fdb4f5120b35b081f3fc4bdc</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ecf0a1496f32eaf34825b3d11432c6b1d03c951</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 942/942 (100.0%)</li>
<li>Failed: 0/942 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 2155/2156 (100.0%)</li>
<li><details><summary>Failed: 1/2156 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.object_management.multithreaded_shared_resources.compute_pipeline
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>


</ul></details>

File: llpc/tool/amdllpc.cpp [hunk 1]
```diff
@@ -452,54 +452,72 @@ static Result Init(
             }
         }
 
-        // Initialize the path for shader cache
-        constexpr uint32_t MaxFilePathLen = 512;
-        char               shaderCacheFileDirOption[MaxFilePathLen];
+        const char* pName = "-shader-cache-file-dir";
+        size_t nameLen = strlen(pName);
```
Comment (kuhar) [inline llpc/tool/amdllpc.cpp]: nit: `const size_t nameLen`

File: llpc/tool/amdllpc.cpp [hunk 2]
```diff
@@ -452,54 +452,72 @@ static Result Init(
...
+        bool found = false;
+        const char* pArg = nullptr;
```
Comment (kuhar) [inline llpc/tool/amdllpc.cpp]: this can be moved inside the loop

File: llpc/tool/amdllpc.cpp [hunk 3]
```diff
@@ -452,54 +452,72 @@ static Result Init(
...
+        for (int32_t i = 1; i < argc; ++i)
+        {
+            pArg = argv[i];
+            if ((strncmp(pArg, pName, nameLen) == 0) &&
```
Comment (kuhar) [inline llpc/tool/amdllpc.cpp]: you can declare `name` as `llvm::StringLiteral` and use make `pArg` be `llvm::StringRef args` and check if if starts with what you expect by calling `arg.starts_with`

File: llpc/tool/amdllpc.cpp [hunk 4]
```diff
@@ -452,54 +452,72 @@ static Result Init(
...
+        if (found == false)
+        {
+            // Initialize the path for shader cache
+            constexpr uint32_t MaxFilePathLen = 512;
+            char               shaderCacheFileDirOption[MaxFilePathLen];
+
+            // Initialize the root path of cache files
+            // Steps:
+            //   1. Find AMD_SHADER_DISK_CACHE_PATH to keep backward compatibility.
+            const char* pEnvString = getenv("AMD_SHADER_DISK_CACHE_PATH");
```
Comment (kuhar) [inline llpc/tool/amdllpc.cpp]: llvm has functions that deal with env: https://llvm.org/doxygen/classllvm_1_1sys_1_1Process.html. 
`llvm::GetEnv` returns an optional string with the value. Can you use that instead?
If this is not consistent with the rest of the tool, this can be a separate PR.

File: llpc/tool/amdllpc.cpp [hunk 5]
```diff
@@ -452,54 +452,72 @@ static Result Init(
...
 #ifdef WIN_OS
-        //   2. Find LOCALAPPDATA.
-        if (pEnvString == nullptr)
-        {
-            pEnvString = getenv("LOCALAPPDATA");
-        }
+            //   2. Find LOCALAPPDATA.
+            if (pEnvString == nullptr)
+            {
+                pEnvString = getenv("LOCALAPPDATA");
```
Comment (kuhar) [inline llpc/tool/amdllpc.cpp]: same for this and the ones below

File: llpc/tool/amdllpc.cpp [hunk 6]
```diff
@@ -452,54 +452,72 @@ static Result Init(
...
-            if (pEnvString != nullptr)
+            if (pEnvString == nullptr)
             {
-                snprintf(shaderCacheFileRootDir, sizeof(shaderCacheFileRootDir),
-                    "%s/.cache", pEnvString);
-                pEnvString = &shaderCacheFileRootDir[0];
+                pEnvString = getenv("HOME");
+                if (pEnvString != nullptr)
+                {
+                    snprintf(shaderCacheFileRootDir, sizeof(shaderCacheFileRootDir),
```
Comment (kuhar) [inline llpc/tool/amdllpc.cpp]: Why isn't shaderCacheFileRootDir a `std::string`?

File: llpc/tool/amdllpc.cpp [hunk 7]
```diff
@@ -452,54 +452,72 @@ static Result Init(
...
-        {
-            snprintf(shaderCacheFileDirOption, sizeof(shaderCacheFileDirOption),
-                     "-shader-cache-file-dir=%s", pEnvString);
-        }
-        else
-        {
-            strncpy(shaderCacheFileDirOption, "-shader-cache-file-dir=.", sizeof(shaderCacheFileDirOption));
+            if (pEnvString != nullptr)
+            {
+                snprintf(shaderCacheFileDirOption, sizeof(shaderCacheFileDirOption),
```
Comment (kuhar) [inline llpc/tool/amdllpc.cpp]: same

**Comment by jayfoad** [comment]
@kuhar I can't argue with any of your suggestions, but all I did in this commit is copy the "loop over argv" code from earlier in the same file (I realise the duplication is not ideal) and re-indent the "query environment variables" code.

**Comment by kuhar** [comment]
Fair enough, let's keep these things separate.

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit d9cd7189824f2ec2ecfe747e55bf1768266755c4
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 9fab16015e522fff05890a045a1e9d8d3c23a636</li>
<li>SPVGEN: dev a61db14f75728e91c243d4bfc1c39700e9dc0ba6</li>
<li>XGL: dev 052ebfa2e77abd14fdb4f5120b35b081f3fc4bdc</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ecf0a1496f32eaf34825b3d11432c6b1d03c951</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 942/942 (100.0%)</li>
<li>Failed: 0/942 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 65463/65463 (100.0%)</li>
<li>Failed: 0/65463 (0.0%)</li>
</ul></details>


</ul></details>
